common/interfaces/card_interface/nfc_events.c

```diff
@@ -0,0 +1,221 @@
+
+/*****************************************************************************
+ * STATIC FUNCTIONS
+ *****************************************************************************/
+static bool check_card_removed_status(uint8_t retry_count) {
+  uint32_t err = adafruit_diagnose_card_presence();
+  if (err != PN532_DIAGNOSE_CARD_DETECTED_RESP) {
+    card_removal_retry_counter++;
+  } else {
+    card_removal_retry_counter = 0;
+  }
+
+  if (card_removal_retry_counter > retry_count) {
+    return true;
+  }
+  return false;
+}
+
+static void nfc_handle_card_select_resp() {
+  nfc_a_tag_info nfc_tag_info;
+  uint32_t card_select_status = pn532_read_nfca_target_init_resp(&nfc_tag_info);
+  if (card_select_status == STM_SUCCESS) {
+    nfc_set_card_detect_event();
+  } else if (card_select_status != NFC_RESP_NOT_READY) {
+    nfc_state = NFC_STATE_SET_SELECT_CARD_CMD;
+  }
+}
+
+/*****************************************************************************
+ * GLOBAL FUNCTIONS
+ *****************************************************************************/
+bool nfc_get_event(nfc_event_t *nfc_event_os_obj) {
+  ASSERT(nfc_event_os_obj != NULL);
+
+  if (nfc_event.event_occured == true) {
+    memcpy(nfc_event_os_obj, &nfc_event, sizeof(nfc_event_t));
+    nfc_reset_event();
+    return true;
+  }
+  return false;
+}
+
+void nfc_reset_event() {
+  memzero(&nfc_event, sizeof(nfc_event_t));
+}
+
+void nfc_set_card_detect_event() {
+  nfc_event.event_occured = true;
+  nfc_event.event_type = NFC_EVENT_CARD_DETECT;
+  nfc_state = NFC_STATE_CARD_DETECTED;
+}
+
+void nfc_set_card_removed_event() {
+  nfc_event.event_occured = true;
+  nfc_event.event_type = NFC_EVENT_CARD_REMOVED;
+  nfc_state = NFC_STATE_CARD_REMOVED;
+}
+
+void nfc_en_select_card_task() {
+  nfc_state = NFC_STATE_SET_SELECT_CARD_CMD;
+}
+
+uint32_t nfc_en_wait_for_card_removal_task() {
+  uint32_t card_presence_state = nfc_diagnose_card_presence();
+  if (card_presence_state == PN532_DIAGNOSE_CARD_DETECTED_RESP) {
+    nfc_state = NFC_STATE_WAIT_FOR_CARD_REMOVAL;
+    card_removal_retry_counter = 0;
+  }
+  return card_presence_state;
+}
+
+void nfc_task_handler() {
+  switch (nfc_state) {
+    case NFC_STATE_SET_SELECT_CARD_CMD:
+      if (pn532_set_nfca_target_init_command() == STM_SUCCESS)
+        nfc_state = NFC_STATE_WAIT_SELECT_CARD_RESP;
+      break;
+
+    case NFC_STATE_WAIT_SELECT_CARD_RESP: {
+      nfc_handle_card_select_resp();
+    } break;
+
+    case NFC_STATE_CARD_DETECTED: {
+      // Should never reach here.
+    } break;
+
+    case NFC_STATE_WAIT_FOR_CARD_REMOVAL: {
+      if (check_card_removed_status(DEFAULT_CARD_REMOVAL_RETRY_COUNT)) {
+        nfc_set_card_removed_event();
+      }
+    } break;
+
+    case NFC_STATE_CARD_REMOVED: {
+      // Should never reach here.
+    } break;
+
+    default:
+      LOG_ERROR("xxx37: %d", nfc_state);
+      break;
+  }
+}
+
+void nfc_ctx_destroy() {
+  if (nfc_state != NFC_STATE_OFF) {
```

Comment:
This check seems to be redundant as it's guaranteed that `nfc_ctx_destroy` is called once per call to `get_event`

Reply:
not per call to `get_event`, only when nfc event is enabled in `get_event`.
Will remove this check though.

Reply:
added